rfcs/003 - Multi-Region Support RFC.md

```diff
@@ -0,0 +1,137 @@
+
+##### Cons
+
+1. TypeORM entity logic needs trickery to take a single logical entity and map it to several similarly-named entities, and then somehow inform the mapper of what trickery was used for which instance.
+2. You can't hide regions you don't care about, and every query has to be altered.
+
+#### `Region` mega-join table
+
+A specialized join table that contains `table`, `column`, `value`, and `region` columns to uniquely identify *any* row of any table and specify the region that applies to it. For rows that don't have a record in there, a `default_region` column in the `aws_account` would be assumed to be the answer if it is relevant. There could also be a list of `active_regions` that can trim the AWS API access to only regions the user is interested in.
+
+##### Pros
+
+1. Also requires zero schema changes to other modules, can be done in a single shot.
+2. Preserves the "default region" concept to keep inserts effectively identical.
+3. Can actually be faster by cutting out irrelevant regions.
+
+##### Cons
+
+1. Joins would be very slow and meta, with no good way to guarantee correctness as the joins will not be enforced by Postgres, but only us.
+2. Inserts for non-default regions are all two-step, and *must* be explicitly done before `iasql_apply` or it will accidentally create in the wrong region.
+3. Querying what's in what region will be much tougher than other options.
+
+#### Region-prefixed-to-@cloudId-column
+
+The IaSQL `@cloudId` concept (an ID that isn't the autogenerated database ID, but the ID the cloud provider generates once the resource is created) can double-up as the region indicator, when necessary. On insert you insert the region in question, and once it is created in the cloud, it it set to `region:cloudId` in the column.
+
+##### Pros
+
+1. This also requires no schema changes and can be updated over time.
+2. Could similarly have a `default_region` and `active_regions` concept like the mega-join table to gain its benefits.
+
+##### Cons
+
+1. Querying the records by region will be complicated with string prefix checking, so can never be performant, and also varies entity-by-entity.
+2. The cloud ID itself similarly would need to be parsed out to find the relevant part of the string, and so joins between entities *on* the cloudId become weirder.
+
+#### Optional region column on relevant tables
+
+Coming full circle on this, we can improve on the first option in this list by taking the `default_region` and `active_regions` idea from the prior two options and applying that to it.
+
+##### Pros
+
+1. Inserts *can* stay the same and not specify a region, with the mappers automatically assuming the default region in question.
+2. It can also be migrated to over time.
+
+##### Cons
+
+1. Nulling-out the region column after it has been populated can have unpredictable results, and is easier to accidentally do than the prior option.
```

Comment:
This is the option I had thought about most recently after we talked on Friday. In my head the `region` column in `aws_account` would be renamed to `default_region` and still be mandatory, but fall back to `us-east-1` as the default. The `region` column in every region-specific entity could then be mandatory which would prevent this con? Unless I am missing something

Reply:
So if you require the `region` column to be defined, then you don't need the `default_region` property as it wouldn't do anything?

Reply:
Also, requiring that column then alters all `insert`s to require specification of the `region` column, which restores the same issue as in the first option listed.

Reply:
I guess my thought was that every `region` column would need to be defined and would fall back to a default which is the foreign key relation to whatever the `default_region` is. However, I think your point might be what happens if the `default_region` changes after it is set?  I think we could add a trigger that updates every single `region` field?

Reply:
Talked offline. Part of my confusion was that the default to the `region` column on every table is a schema change. However, we came up with yet another option that lets the default of the `region` column be defined by a function that simply looks at the `default_region` field in `aws_account`